pkg/disk/partition_table.go

```diff
@@ -160,6 +161,146 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	// directory requirements are additive, meaning the minimum size for a
+	// mountpoint's partition is the sum of all the required directory sizes it
+	// will contain.
+	RequiredMinSizes map[string]uint64
+}
+
+// Returns the default filesystem type if the fstype is empty. If both are
+// empty/none, returns an error.
+func (options *CustomPartitionTableOptions) getfstype(fstype string) (string, error) {
+	if fstype != "" {
+		return fstype, nil
+	}
+
+	if options.DefaultFSType == FS_NONE {
+		return "", fmt.Errorf("no filesystem type defined and no default set")
+	}
+
+	return options.DefaultFSType.String(), nil
+}
+
+// NewCustomPartitionTable creates a partition table based almost entirely on the partitioning customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
+	if options == nil {
+		// init options with empty struct to use defaults
+		options = &CustomPartitionTableOptions{
+			PartitionTableType: PT_GPT,
+		}
+	}
+
+	if customizations == nil {
+		customizations = &blueprint.PartitioningCustomization{}
+	}
+
+	// validate the partitioning customizations before using them
+	if err := customizations.Validate(); err != nil {
+		return nil, fmt.Errorf("error generating partition table: %w", err)
+	}
+
+	pt := &PartitionTable{}
+
+	switch options.PartitionTableType {
+	case PT_GPT, PT_DOS:
+		pt.Type = options.PartitionTableType.String()
+	case PT_NONE:
+		// default to "gpt"
+		pt.Type = PT_GPT.String()
+	default:
+		return nil, fmt.Errorf("error generating partition table: invalid partition table type enum value: %d", options.PartitionTableType)
+	}
+
+	// The boot type will be the default only if it's a supported filesystem
+	// type for /boot (ext4 or xfs). Otherwise, we default to xfs.
+	// FS_NONE also falls back to xfs.
+	var bootType FSType
+	switch options.DefaultFSType {
+	case FS_EXT4, FS_XFS:
+		bootType = options.DefaultFSType
+	default:
+		bootType = FS_XFS
+	}
+
+	if err := addPartitionsForBootMode(pt, options.BootMode); err != nil {
+		return nil, fmt.Errorf("error generating partition table: %w", err)
+	}
+	if customizations.LVM != nil || customizations.Btrfs != nil {
+		// we need a /boot partition to boot LVM or Btrfs, create boot
+		// partition if it does not already exist
+		ensureBoot(pt, bootType)
+	}
+
+	if err := addPartitionsForPlain(pt, customizations.Plain, options); err != nil {
+		return nil, fmt.Errorf("error generating partition table: %w", err)
+	}
+	if err := addPartitionsForLVM(pt, customizations.LVM, options); err != nil {
+		return nil, fmt.Errorf("error generating partition table: %w", err)
+	}
+	addPartitionsForBtrfs(pt, customizations.Btrfs)
+
+	if err := ensureRoot(pt, options); err != nil {
+		return nil, err
+	}
+
+	if requiredSizes := options.RequiredMinSizes; len(requiredSizes) != 0 {
```

Comment:
(super nitpick) not sure we need the temporary assignment here (or am I missing something?), maybe just:
```go
diff --git a/pkg/disk/partition_table.go b/pkg/disk/partition_table.go
index 8a682668e..36de2f707 100644
--- a/pkg/disk/partition_table.go
+++ b/pkg/disk/partition_table.go
@@ -291,8 +292,8 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
                return nil, err
        }
 
-       if requiredSizes := options.RequiredMinSizes; len(requiredSizes) != 0 {
-               pt.EnsureDirectorySizes(requiredSizes)
+       if len(options.RequiredMinSizes) != 0 {
+               pt.EnsureDirectorySizes(options.RequiredMinSizes)
        }
 
        pt.relayout(customizations.MinSize)
```

Reply:
Done.